### PR TITLE
feat(symphony): add rolling TPS sparkline to TUI dashboard

### PR DESCRIPTION
**Pull request by gannonh**
## Summary
- add a `ThroughputTracker` in Symphony TUI to sample cumulative token totals on each 500ms refresh
- compute current TPS using a rolling 5-second window and compute a 24-bucket sparkline across a 10-minute window
- render a new `Throughput: <tps> <sparkline>` row in the TUI summary area using `▁▂▃▄▅▆▇█`
- add TUI unit tests for TPS math, zero-throughput flattening, recent-spike bucket rendering, sample retention trimming, and rendered-row presence

## Testing
- `cd apps/symphony && cargo test`
- `cd apps/symphony && cargo clippy -- -D warnings`

## Ticket
- KAT-893

## Summary by CodeRabbit

* **New Features**
  * Real-time throughput tracking showing tokens processed per second on the dashboard
  * Sparkline visualization displaying throughput trends over a rolling 10-minute window
  * Expanded dashboard summary section to include throughput metrics alongside job status information

<h3>Greptile Summary</h3>

This PR adds a rolling throughput sparkline row to the Symphony TUI dashboard by introducing a `ThroughputTracker` struct that samples cumulative token totals every 500 ms, computes a 5-second rolling TPS, and distributes token counts into 24 time-proportional buckets across a 10-minute sparkline window rendered using Unicode block characters.

**Key changes:**
- New `ThroughputTracker` struct with `record_sample`, `current_tps`, `sparkline`, `bucket_tps`, `trim_old_samples`, and `window_token_delta` methods — all logic is self-contained within `tui.rs`
- Token counter resets (decreasing `total_tokens`) are handled by clearing all samples and starting fresh
- Out-of-order and duplicate timestamps are handled gracefully in `record_sample`
- `draw_dashboard` receives a pre-computed `throughput_line: &str` (computed once per tick in `run_tui`) and renders it as a second line in the summary section; `Constraint::Length` for the summary area is widened from 1 to 2 rows accordingly
- Five focused unit tests cover TPS math, zero-throughput flattening, last-bucket spike rendering, sample-retention trimming, and full dashboard rendering via `TestBackend`

**Observations:**
- `current_tps` always divides by the full 5-second window denominator, so TPS is systematically underreported during the first ~5 s of a run (e.g., 2.0 tps displayed when the true rate is 20.0 tps in the first 500 ms tick). This is a known characteristic of rolling-window metrics and appears intentional, but is undocumented.
- Out-of-order timestamps are silently dropped with no log emit; a `tracing::warn!` would aid debugging during clock-skew events.

<h3>Confidence Score: 4/5</h3>

- Safe to merge — logic is correct and well-tested; only minor design documentation gaps identified.
- The implementation is careful: edge cases (counter resets, duplicate/out-of-order timestamps, empty windows) are all handled, the bucket distribution math is correct, the index arithmetic in `bucket_tps` is bounds-safe, and five targeted unit tests confirm the key behaviours. The only actionable items are a missing comment documenting the intentional cold-start TPS underestimation and the absence of a warn log when samples are silently dropped due to backwards clock — neither affects correctness.
- No files require special attention.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| apps/symphony/src/tui.rs | Adds ThroughputTracker struct with time-proportional bucket distribution, sparkline rendering, and TPS computation; integrates into run_tui loop and draw_dashboard. Logic is sound and well-tested. Minor design notes around cold-start TPS underestimation and silent out-of-order sample dropping. |

</details>

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Orchestrator
    participant SnapshotHandle
    participant run_tui
    participant ThroughputTracker
    participant draw_dashboard
    participant Terminal

    loop Every 500 ms (REFRESH_INTERVAL_MS)
        run_tui->>SnapshotHandle: snapshot_handle.read()
        SnapshotHandle-->>run_tui: OrchestratorSnapshot { codex_totals.total_tokens }
        run_tui->>run_tui: now_ms = Utc::now().timestamp_millis()
        run_tui->>ThroughputTracker: record_sample(now_ms, total_tokens)
        Note over ThroughputTracker: Guard: drop if out-of-order<br/>Max if duplicate ts<br/>Clear if token reset
        ThroughputTracker->>ThroughputTracker: trim_old_samples(now_ms)
        run_tui->>ThroughputTracker: throughput_line(now_ms)
        ThroughputTracker->>ThroughputTracker: current_tps(now_ms)<br/>[window_token_delta / 5s]
        ThroughputTracker->>ThroughputTracker: sparkline(now_ms)<br/>[bucket_tps → normalise → SPARKLINE_BLOCKS]
        ThroughputTracker-->>run_tui: "Throughput: X.X tps ▁▂▃▄▅▆▇█..."
        run_tui->>draw_dashboard: frame, snapshot, now, throughput_line
        draw_dashboard->>Terminal: render Paragraph([summary_line, throughput_line])
    end
```

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: apps/symphony/src/tui.rs
Line: 61-73

Comment:
**Cold-start TPS underestimation**

`current_tps` always divides by the full `CURRENT_TPS_WINDOW_MS` (5 seconds) regardless of how much historical data is actually available. At startup — e.g., after just the first two 500 ms ticks — the display will show roughly 1/10th the true rate. For example, 10 tokens in the first 500 ms yields `10 / 5.0 = 2.0 tps` even though the instantaneous rate is `20.0 tps`.

This is a common trade-off for rolling-window metrics and is consistent with the PR description ("rolling 5-second window"), so this is likely intentional. Consider a brief comment to document the expected cold-start behaviour so future maintainers don't mistake the early-seconds underreport for a bug:

```rust
// NOTE: we always divide by the full window length, even at startup when less
// than CURRENT_TPS_WINDOW_MS of data is available. This means TPS will be
// under-reported during the first ~5 s but converges to the true rate once
// the window is fully populated.
fn current_tps(&self, now_ms: i64) -> f64 {
```

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: apps/symphony/src/tui.rs
Line: 36-53

Comment:
**Silent sample drop on backwards clock**

When `timestamp_ms < *last_ts`, `record_sample` silently returns without recording or logging anything. If the system clock steps backward (NTP correction, VM migration, etc.) the tracker will quietly stall — emitting the same stale sparkline and TPS — until `Utc::now()` catches back up to the last recorded timestamp. For a dashboard refresh running every 500 ms this is unlikely to be noticeable, but a `tracing::warn!` on the out-of-order branch would make such a regression much easier to diagnose:

```rust
Some((last_ts, _last_total)) if timestamp_ms < *last_ts => {
    tracing::warn!(
        last_ts = *last_ts,
        received_ts = timestamp_ms,
        "throughput tracker received out-of-order timestamp; skipping sample"
    );
    return;
}
```

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Reviews (1): Last reviewed commit: ["fix(symphony): resolve clippy repeat_n l..."](https://github.com/gannonh/kata/commit/882ea7046cdf1d53260a322700648b5242c6848f) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=25981386)</sub>

> Greptile also left **2 inline comments** on this PR.

### CODE REVIEW

File: apps/symphony/src/tui.rs [hunk 1]
```diff
@@ -21,6 +21,183 @@ use crate::domain::OrchestratorSnapshot;
...
+        }
+        let window_start = now_ms.saturating_sub(CURRENT_TPS_WINDOW_MS);
+        let delta = self.window_token_delta(window_start, now_ms);
+        // Intentionally divide by the fixed 5s window so the dashboard readout
+        // stays stable; this smooths noise but under-reports during warm-up.
+        let window_seconds = (CURRENT_TPS_WINDOW_MS as f64) / 1_000.0;
+        if window_seconds > 0.0 {
+            delta / window_seconds
+        } else {
+            0.0
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/symphony/src/tui.rs:78]: Addressed in `9b79d8a`: added an inline note in `current_tps` documenting fixed-window smoothing and cold-start under-reporting behavior.

File: apps/symphony/src/tui.rs [hunk 2]
```diff
@@ -21,6 +21,183 @@ use crate::domain::OrchestratorSnapshot;
...
+            }
+            Some((_, last_total)) if total_tokens < *last_total => {
+                self.token_samples.clear();
+            }
+            _ => {}
+        }
+
+        self.token_samples.push((timestamp_ms, total_tokens));
+        self.trim_old_samples(timestamp_ms);
+    }
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/symphony/src/tui.rs:58]: Addressed in `8b027e1`: `record_sample` now emits `tracing::warn!` when receiving out-of-order timestamps before skipping the sample.
